File: modules/snoozepulse-audio/ios/SnoozePulseAudioModule.swift
import ExpoModulesCore
import Foundation

/**
 * Expo bridge for iOS capture (Task 5.2). Same event names / method surface as Android.
 * Swift is written for later macOS compile/verify — not runtime-validated on Windows (ADR-17).
 *
 * Task 6.3 hands the warmed `SnoreClassifier` to `CaptureEngine` so the capture path
 * drives detection through YAMNet (ADR-21, ADR-23). When warmup fails, the engine is
 * still constructed with a nil classifier; the pipeline degrades to "no episodes emitted"
 * rather than crashing capture.
 */
public class SnoozePulseAudioModule: Module {
  private var engine: CaptureEngine?
  private var classifier: SnoreClassifier?
  private var lastBaselineDb: Double?

  private func ensureEngine() -> CaptureEngine {
    if let engine { return engine }
    warmClassifierIfNeeded()
    let capture = CaptureEngine(classifier: classifier)
    capture.onLevel = { [weak self] payload in
      self?.sendEvent("onAudioLevel", payload)
    }
    capture.onSnore = { [weak self] payload in
      self?.sendEvent("onSnore", payload)
    }
    capture.onInterruption = { [weak self] payload in
      self?.sendEvent("onInterruption", payload)
    }
    engine = capture
    return capture
  }

  private func warmClassifierIfNeeded() {
    if classifier != nil { return }
    let created = YamnetClassifier()
    do {
      let delegate = try created.warm()
      NSLog("SnoreClassifier ready (delegate=%@)", delegate)
      classifier = created
    } catch {
      NSLog("SnoreClassifier warmup failed; running without ML detector: %@",
            String(describing: error))
      created.close()
    }
  }

  public func definition() -> ModuleDefinition {
    Name("SnoozePulseAudio")

    Events("onAudioLevel", "onSnore", "onInterruption")

    Constant("HEARTBEAT_INTERVAL_MS") {
      150
    }

    OnDestroy {
      self.engine?.stop()
      self.engine = nil
      self.classifier?.close()
      self.classifier = nil
    }

    AsyncFunction("startRecording") { (id: String, promise: Promise) in
      do {
        try self.ensureEngine().start(sessionId: id, baselineDb: self.lastBaselineDb)
        promise.resolve(nil)
      } catch {
        promise.reject("AUDIO_BUSY", error.localizedDescription)
      }
    }

    AsyncFunction("stopRecording") { (promise: Promise) in
      self.engine?.stop()
      promise.resolve(nil)
    }

    AsyncFunction("pauseRecording") { (promise: Promise) in
      guard let capture = self.engine, capture.running() else {
        promise.reject("AUDIO_ENGINE", "Not recording")
        return
      }
      capture.pause()
      promise.resolve(nil)
    }

    AsyncFunction("resumeRecording") { (promise: Promise) in
      guard let capture = self.engine, capture.running() else {
        promise.reject("AUDIO_ENGINE", "Not paused")
        return
      }
      capture.resume()
      promise.resolve(nil)
    }

    AsyncFunction("calibrate") { (promise: Promise) in
      do {
        let result = try self.ensureEngine().calibrate()
        self.lastBaselineDb = result["baselineDb"] as? Double
        promise.resolve(result)
      } catch {
        promise.reject("CALIBRATION", error.localizedDescription)
      }
    }
  }
}

File: modules/snoozepulse-audio/android/src/main/java/expo/modules/snoozepulseaudio/SnoozePulseAudioModule.kt
package expo.modules.snoozepulseaudio

import android.util.Log
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo bridge for the native capture engine (Task 5.2).
 * JS receives only throttled levels / snore / interruption events — never PCM.
 *
 * Task 6.3 hands the warmed [SnoreClassifier] to [CaptureEngine] so the capture path
 * drives detection through YAMNet (ADR-21, ADR-23). When warmup fails, the engine is
 * still constructed with a null classifier; the pipeline degrades to "no episodes emitted"
 * rather than crashing capture.
 */
class SnoozePulseAudioModule : Module() {
  companion object {
    private const val TAG = "SnoozePulseAudioModule"
  }

  private var engine: CaptureEngine? = null
  private var classifier: SnoreClassifier? = null
  /** Last calibration baseline, applied on the next [startRecording]. */
  private var lastBaselineDb: Double? = null

  private fun ensureEngine(): CaptureEngine {
    engine?.let { return it }
    val ctx =
      appContext.reactContext
        ?: throw IllegalStateException("React context not ready")
    warmClassifierIfNeeded(ctx)
    val created =
      CaptureEngine(
        context = ctx,
        classifier = classifier,
        emitLevel = { payload -> sendEvent("onAudioLevel", payload) },
        emitSnore = { payload -> sendEvent("onSnore", payload) },
        emitInterruption = { payload -> sendEvent("onInterruption", payload) },
      )
    engine = created
    return created
  }

  private fun warmClassifierIfNeeded(ctx: android.content.Context) {
    if (classifier != null) return
    val created = YamnetClassifier(ctx)
    try {
      val delegate = created.warm()
      Log.i(TAG, "SnoreClassifier ready (delegate=$delegate)")
      classifier = created
    } catch (error: Exception) {
      Log.e(TAG, "SnoreClassifier warmup failed; running without ML detector", error)
      created.close()
    }
  }

  override fun definition() = ModuleDefinition {
    Name("SnoozePulseAudio")

    Events("onAudioLevel", "onSnore", "onInterruption")

    Constant("HEARTBEAT_INTERVAL_MS") {
      150
    }

    OnDestroy {
      engine?.stop()
      engine = null
      classifier?.close()
      classifier = null
    }

    AsyncFunction("startRecording") { id: String, promise: Promise ->
      try {
        ensureEngine().start(id, lastBaselineDb)
        promise.resolve(null)
      } catch (error: IllegalStateException) {
        promise.reject("AUDIO_BUSY", error.message, error)
      } catch (error: SecurityException) {
        promise.reject("PERMISSION_DENIED", error.message, error)
      } catch (error: Exception) {
        promise.reject("AUDIO_ENGINE", error.message, error)
      }
    }

    AsyncFunction("stopRecording") { promise: Promise ->
      engine?.stop()
      promise.resolve(null)
    }

    AsyncFunction("pauseRecording") { promise: Promise ->
      val capture = engine
      if (capture == null || !capture.isRunning()) {
        promise.reject("AUDIO_ENGINE", "Not recording", null)
        return@AsyncFunction
      }
      capture.pause()
      promise.resolve(null)
    }

    AsyncFunction("resumeRecording") { promise: Promise ->
      val capture = engine
      if (capture == null || !capture.isRunning()) {
        promise.reject("AUDIO_ENGINE", "Not paused", null)
        return@AsyncFunction
      }
      capture.resume()
      promise.resolve(null)
    }

    AsyncFunction("calibrate") { promise: Promise ->
      try {
        val result = ensureEngine().calibrate()
        lastBaselineDb = result["baselineDb"] as? Double
        promise.resolve(result)
      } catch (error: Exception) {
        promise.reject("CALIBRATION", error.message, error)
      }
    }
  }
}
